### ISSUE TITLE
Sik-Ho Tsang | Review: Representation Learning with Contrastive Predictive Coding (CPC/CPCv1).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#71]
Sik-Ho Tsang. [Review: Representation Learning with Contrastive Predictive Coding (CPC/CPCv1)](https://sh-tsang.medium.com/review-representation-learning-with-contrastive-predictive-coding-cpc-cpcv1-8ec7c6e91a8e).

**Comment by NorbertZheng** [comment]
## Overview
Representation Learning Using InfoNCE Loss. In this story, Representation Learning with Contrastive Predictive Coding, (CPC/CPCv1), by DeepMind, is reviewed. In this paper:
 - **Contrastive Predictive Coding (CPC)** is proposed, which is a universal unsupervised learning approach to extract useful representations from high-dimensional data.
 - A **noise-Contrastive Estimation Loss**, namely InfoNCE Loss, is used, which induces the latent space to capture information that is maximally useful to predict future samples.
 - Four distinct domains are used for evaluation: speech, images, text and reinforcement learning in 3D environments.

This is a paper in 2018 arXiv with over 1800 citations. This paper makes use of [NCE](https://sh-tsang.medium.com/review-learning-word-embeddings-efficiently-with-noise-contrastive-estimation-nce-dba1c345c153) and [Negative Sampling](https://sh-tsang.medium.com/review-distributed-representations-of-words-and-phrases-and-their-compositionality-negative-ab9ebfc3f041) in NLP for representation learning/self-supervised learning.

**Comment by NorbertZheng** [comment]
## Motivation and Intuition of Contrastive Predictive Coding (CPC)
 - ~~**Unimodal losses** such as means squared error and cross-entropy~~ are not very useful, modeling $p(x|c)$ directly may not be optimal, where target $x$ (future) and context $c$ (present).
 - For example, images may contain thousands of bits of information while the high-level latent variables such as the class label contain much less information (10 bits for 1,024 categories).
 - However, unsupervised learning is yet to see a breakthrough as well.
 - **Predictive coding has been used for long time in data compression.**

 - The main intuition of the model is to learn the representations that encode the underlying shared information between different parts of the (high-dimensional) signal. At the same time it discards low-level information and noise that is more local.
 - When predicting future information, CPC instead encode the target $x$ (future) and context $c$ (present) into a compact distributed vector representations in a way that maximally preserves the mutual information of the original signals $x$ and $c$ defined as:

![image](https://user-images.githubusercontent.com/37235023/221484193-4efdcef6-ecb6-40cb-84c4-423466bcbc09.png)

**Comment by NorbertZheng** [comment]
## Contrastive Predictive Coding (CPC): Overview
![image](https://user-images.githubusercontent.com/37235023/221742071-a7de854a-e65e-4017-a8aa-4bae6936f431.png)
 - $g_{enc}$: First, a non-linear encoder $g_{enc}$ maps the input sequence of observations $x_{t}$ to a sequence of latent representations $z_{t}=g_{enc}(x_{t})$, potentially with a lower temporal resolution.
 - $g_{ar}$: Next, an autoregressive model $g_{ar}$ summarizes all $z_{\leq t}$ in the latent space and produces a context latent representation $c_{t}=g_{ar}(z_{\leq t})$.

As argued in the previous section, we do not predict future observations $x_{t+k}$ directly with a generative model $f_{k}\propto (x_{t+k}|c_{t})$. Instead we model a density ratio which preserves the mutual information between $x_{t+k}$ and $c_{t}$:

![image](https://user-images.githubusercontent.com/37235023/221485567-b287df72-2deb-4812-a016-86827a9c059c.png)

 - The density ratio $f$ can be unnormalized.
 - A simple log-bilinear model can be used:

![image](https://user-images.githubusercontent.com/37235023/221485918-94fcd570-396c-4f13-8cdd-6b431968b232.png)

A linear transformation $W_{k}^{T}c_{t}$ is used for the prediction with a different $W_{k}$ for every step $k$.
 - Although $p(x)$ or $p(x|c)$ cannot be evaluated directly, samples from these distributions can be utilized such as by Noise-Contrastive Estimation (NCE).

In the proposed model, either of $z_{t}$ and $c_{t}$ could be used as representation for downstream tasks.

**Comment by NorbertZheng** [comment]
## InfoNCE Loss and Mutual Information Estimation
 - Both the encoder and autoregressive model are trained to jointly optimize a loss based on NCE, called InfoNCE.
 - Given a set $X=\\{x_{1},...,x_{N}\\}$ of $N$ random samples containing one positive sample from $p(x_{t+k}|c_{t})$ and $N-1$ negative samples from the "proposal" distribution $p(x_{t+k})$.
![image](https://user-images.githubusercontent.com/37235023/221742665-681bc17e-fb4b-4bbd-b6ec-698d7c05bd2f.png)

 - Recall in the previous section that:
![image](https://user-images.githubusercontent.com/37235023/221742866-57f0a952-e628-4db1-8530-91e12f36ecc1.png)

 - Optimizing this loss will result in $f_{k}(x_{t+l},c_{t})$ estimating the density ratio mentioned in the previous section.
 - The optimal probability for this loss is written as $p(d=i|X,c_{t})$ with $[d=i]$ being the indicator that sample $x_{i}$ is the "positive" sample.
 - The probability that sample $x_{i}$ was drawn from the conditional distribution $p(x_{t+k}|c_{t})$ rather than the proposal distribution $p(x_{t+k})$ can be derived as follows:
![image](https://user-images.githubusercontent.com/37235023/221743836-854d1b65-090e-4c5b-bc0e-fe9c5acbe254.png)

 - As seen, the optimal value for $f_{k}(x_{t+k},c_{t})$ is independent of the the choice of the number of negative samples $N-1$.
 - Minimizing the InfoNCE loss $L_{n}$, is actually maximizing a lower bound on the mutual information $I(x_{t+k}, c_{t})$:
![image](https://user-images.githubusercontent.com/37235023/221753239-d2e54bed-c4a2-4161-a11e-3aaab1d9f85c.png)

 - The proof (By splitting $X$ into the positive example and the negative examples $X_{neg}$):
![image](https://user-images.githubusercontent.com/37235023/221753432-24eb7d35-6d87-467d-9283-92044def2184.png)

 - (InfoNCE loss is highly related to [NCE](https://sh-tsang.medium.com/review-learning-word-embeddings-efficiently-with-noise-contrastive-estimation-nce-dba1c345c153) & [Negative Sampling](https://sh-tsang.medium.com/review-distributed-representations-of-words-and-phrases-and-their-compositionality-negative-ab9ebfc3f041) used in NLP. Please feel free to read it if interested.)

**Comment by NorbertZheng** [comment]
## Experiments for Audio
### Pretext Task
 - A 100-hour subset of the publicly available LibriSpeech dataset is used which does not have labels, but only raw text.
 - Authors have made the aligned phone labels and our train/test split. The dataset contains speech from 251 different speakers.
 - $g_{enc}$: The encoder architecture $g_{enc}$ used in our experiments consists of a strided CNN that runs directly on the 16KHz PCM audio waveform. Five convolutional layers with strides [5, 4, 2, 2, 2], filter-sizes [10, 8, 4, 4, 4] and 512 hidden units with ReLU activations, are used.
   - There is a feature vector for **every 10ms of speech** (no overlap!!!).
 - $g_{ar}$: A [GRU](https://sh-tsang.medium.com/review-empirical-evaluation-of-gated-recurrent-neural-networks-on-sequence-modeling-gru-2adb86559257) RNN for the autoregressive part of the model, $g_{ar}$ with 256 dimensional hidden state. The output of the [GRU](https://sh-tsang.medium.com/review-empirical-evaluation-of-gated-recurrent-neural-networks-on-sequence-modeling-gru-2adb86559257) at every timestep is used as the context $c$.
 - A minibatch of 8 examples from which the negative samples in the contrastive loss are drawn.
 - 12 timesteps in the future are predicted using the contrastive loss.
![image](https://user-images.githubusercontent.com/37235023/221755469-6d36cdef-bf10-4a70-a5aa-5ebdca2e842d.png)
Average accuracy of predicting the positive sample in the contrastive loss for 1 to 20 latent steps in the future of a speech waveform.
 - The prediction task becomes harder as the target is further away.

### Downstream Task
![image](https://user-images.githubusercontent.com/37235023/221755955-6954c808-cc54-4e81-9190-6a5d4c63b008.png)
LibriSpeech phone and speaker classification results. For phone classification there are 41 possible classes and for speaker classification 251.

 - To understand the representations extracted by CPC, the phone prediction performance is measured with a linear classifier trained on top of these features.
 - The outputs of the [GRU](https://sh-tsang.medium.com/review-empirical-evaluation-of-gated-recurrent-neural-networks-on-sequence-modeling-gru-2adb86559257) (256 dimensional) are extracted as, $c_{t}$.
 - A multi-class linear logistic regression classifier is trained.
 - Three baselines:
   - Randomly initialized model (i.e., $g_{enc}$ and $g_{ar}$ are untrained),
   - MFCC features,
   - and **a model that is trained end-to-end and supervised with the labeled data**.

For phone classification, CPC obtains 64.6% accuracy. When a single hidden layer is used instead, the accuracy increases from 64.6% to 72.5%, which **is closer to the accuracy of the fully supervised model**.

Interestingly, CPCs capture both speaker identity and speech contents, as demonstrated by the good accuracies attained with a simple linear classifier, which also gets close to the oracle, fully supervised networks.

![image](https://user-images.githubusercontent.com/37235023/221756763-950295a1-023f-4602-91a3-0022e7c43ed7.png)
LibriSpeech phone classification ablation experiments.
 - Some other ablation experiments are also performed, i.e. how many steps to predict, predicting 12 steps obtains the best CPC representations.
 - Also, where the negative samples are drawn from (this part, all are to predcit 12 steps) are also tested. (excl. means excluding negative samples from the current sequence.) It is found that negative samples drawn from the same speaker obtains the best CPC representations.

![image](https://user-images.githubusercontent.com/37235023/221757246-af8949cc-ba61-412e-b726-ed1ff52066d4.png)
t-SNE visualization of audio (speech) representations for a subset of 10 speakers (out of 251).
 - t-SNE visualization shows how discriminative the embeddings are for speaker voice-characteristics.

**Comment by NorbertZheng** [comment]
## Experiments for Vision
### Pretext Task
![image](https://user-images.githubusercontent.com/37235023/221758121-7e2933ea-1121-4b0b-8097-af1f513ff46c.png)
Visualization of Contrastive Predictive Coding for images.

 - $g_{enc}$: A [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------) v2 101 architecture is used as the image encoder $g_{enc}$ to extract CPC representations.
 - After unsupervised training, a linear layer is trained to measure classification accuracy on ImageNet labels.
 - For each of the 64×64 crops, a 60×60 subcrop is picked and padded them back to a 64×64 image. Each crop is then encoded by the [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)-v2–101 encoder. The outputs from the third residual block, are used to spatially mean-pool to get a single 1024-d vector per 64×64 patch.
 - For each image, this results in a 7×7×1024 tensor.
 - $g_{ar}$: A PixelCNN-style autoregressive model (An image generation method) is used to make predictions about the latent activations in following rows top-to-bottom, predict up to five rows from the 7×7 grid.
 - The contrastive loss is applied for each patch in the row.

![image](https://user-images.githubusercontent.com/37235023/221763181-3bad4380-d232-4019-bf1d-e986f42d3cc2.png)
Every row shows image patches that activate a certain neuron in the CPC architecture.

### Downstream Task
![image](https://user-images.githubusercontent.com/37235023/221763503-8cd8d454-1b7c-4ec9-901b-c92eefb738a4.png)
ImageNet top-1 unsupervised classification results.

![image](https://user-images.githubusercontent.com/37235023/221763556-7889d896-e330-4f56-a38d-946e0a0116b2.png)
ImageNet top-5 unsupervised classification results.
 - CPC outperforms [Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff), [Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a), [Jigsaw Puzzles](https://sh-tsang.medium.com/review-unsupervised-learning-of-visual-representations-by-solving-jigsaw-puzzles-50b14d755004), [BiGAN](https://sh-tsang.medium.com/review-bigan-adversarial-feature-learning-gan-535eb76be2ca), etc.
 - Despite being relatively domain agnostic, CPCs improve upon state-of-the-art by 9% absolute in top-1 accuracy, and 4% absolute in top-5 accuracy.

**Comment by NorbertZheng** [comment]
## Experiments for Natural Language
### Pretext Task
 - CPC follows closely the procedure from [Skip-Thought](https://sh-tsang.medium.com/review-skip-thought-vectors-a18565e7255a) vectors model, first learn the unsupervised model on the BookCorpus dataset.
 - (Please feel free to read [Skip-Thought](https://sh-tsang.medium.com/review-skip-thought-vectors-a18565e7255a) vectors if interested.)
 - $g_{enc}$ & $g_{ar}$: The model consists of a simple sentence encoder $g_{enc}$ (a 1D-convolution + ReLU + mean-pooling) that embeds a whole sentence into a 2400-dimension vector $z$, followed by a [GRU](https://sh-tsang.medium.com/review-empirical-evaluation-of-gated-recurrent-neural-networks-on-sequence-modeling-gru-2adb86559257) (2400 hidden units) as $g_{ar}$ which predicts up to 3 future sentence embeddings with the contrastive loss to form $c$.

### Downstream Task
![image](https://user-images.githubusercontent.com/37235023/221766816-601032b6-a556-4859-bd01-29663c32b576.png)
Classification accuracy on five common NLP benchmarks ([40] is [Doc2Vec](https://sh-tsang.medium.com/review-distributed-representations-of-sentences-and-documents-doc2vec-86ef911d4515)).

 - For the classification tasks, the following datasets are used: movie review sentiment (MR) [43], customer product reviews (CR) [44], subjectivity/objectivity [45], opinion polarity (MPQA) [46] and question-type classification (TREC) [47].
 - A logistic regression classifier is trained.

The performance of our method is very similar to the [Skip-Thought](https://sh-tsang.medium.com/review-skip-thought-vectors-a18565e7255a) vector model [26], with the advantage that it does not require a powerful LSTM as word-level decoder, therefore much faster to train.

**Comment by NorbertZheng** [comment]
## Experiments for Reinforcement Learning
![image](https://user-images.githubusercontent.com/37235023/221767105-7cd4d9b3-2851-41aa-a685-39a43335c491.png)
Reinforcement Learning results for 5 DeepMind Lab tasks Black: batched A2C baseline, Red: with auxiliary contrastive loss. 5 reinforcement learning in 3D environments of DeepMind Lab [51] are tested: rooms_watermaze, explore_goal_locations_small, seekavoid_arena_01, lasertag_three_opponents_small and rooms_keys_doors_puzzle. The standard batched A2C [52] agent is used as base model.
 - **CPC is used as an auxiliary loss**, this setting is different from the one in audio, text, and images.
The unroll length for the A2C is 100 steps and we predict up to 30 steps in the future to derive the contrastive loss. For 4 out of the 5 games, performance of the agent improves significantly with the contrastive loss after training on 1 billion frames. (I am not familiar with reinforcement learning, if interested, please feel free to read the paper.)

**Comment by NorbertZheng** [comment]
Later on, CPCv2 is published in 2020 ICLR, hope I have time to review it in the coming future.

**Comment by NorbertZheng** [comment]
## Reference
[2018 arXiv] [CPC/CPCv1]
[Representation Learning with Contrastive Predictive Coding](https://arxiv.org/abs/1807.03748)

**Comment by NorbertZheng** [comment]
## Extended Readings
Self-Supervised Learning
2008–2010 [[Stacked Denoising Autoencoders](https://sh-tsang.medium.com/review-stacked-denoising-autoencoders-self-supervised-learning-c8ff81cef34c)] 2014 [[Exemplar-CNN](https://sh-tsang.medium.com/review-exemplar-cnn-discriminative-unsupervised-feature-learning-with-convolutional-neural-fa68abe937cc)] 2015 [[Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff)] 2016 [[Context Encoders](https://sh-tsang.medium.com/review-context-encoders-feature-learning-by-inpainting-bd181e48997)] [[Colorization](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a)] [[Jigsaw Puzzles](https://sh-tsang.medium.com/review-unsupervised-learning-of-visual-representations-by-solving-jigsaw-puzzles-50b14d755004)] 2017 [[L³-Net](https://sh-tsang.medium.com/review-look-listen-and-learn-self-supervised-learning-ff89a7dee980)] [[Split-Brain Auto](https://sh-tsang.medium.com/review-split-brain-autoencoders-unsupervised-learning-by-cross-channel-prediction-cebeba644a3c)] [[Mean Teacher](https://sh-tsang.medium.com/review-mean-teachers-are-better-role-models-weight-averaged-consistency-targets-improve-b245d5efa5bf)] 2018 [[RotNet/Image Rotations](http://xn--20082010%20[stacked%20denoising%20autoencoders]%202014%20[exemplar-cnn]%202015%20[context%20prediction]%202016%20[context%20encoders]%20[colorization]%20[jigsaw%20puzzles]%202017%20[l3-net][split-brain%20auto]%202018%20[rotnet-f311m/Image%20Rotations])] [[DeepCluster](https://sh-tsang.medium.com/review-deepcluster-deep-clustering-for-unsupervised-learning-of-visual-features-446d0abad74c)] [[CPC/CPCv1](https://sh-tsang.medium.com/review-representation-learning-with-contrastive-predictive-coding-cpc-cpcv1-8ec7c6e91a8e)]
